### ISSUE TITLE
ORDER BY 2 columns does not return same results in 19.9 and 20.8

### ISSUE THREAD

**Issue by astudnev** [ClickHouse/ClickHouse#17960]
I migrating database to 20.8.7.15 and found that. ordering of data in result set appears to be different for the same queries over the same data.

On 19.9 query:

```
SELECT 
    depth, 
    currency_id, 
    SUM(coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.amount) / dictGet('currency', 'divider', toUInt64(1)) AS `_aq.amount`
FROM coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99 
WHERE (coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.depth <= 3) AND (coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.depth > 0)
GROUP BY 
    coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.transfer_from_bin, 
    coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.transfer_to_bin, 
    coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.currency_id, 
    depth
ORDER BY 
    depth ASC, 
    `_aq.amount` DESC
LIMIT 10

┌─depth─┬─currency_id─┬─────────_aq.amount─┐
│     1 │           1 │              16001 │
│     1 │           1 │ 12388.999999999998 │
│     1 │           1 │               8000 │
│     1 │           1 │               7500 │
│     1 │           1 │               7000 │
│     1 │           1 │     4000.099948955 │
│     1 │           1 │               4000 │
│     1 │           1 │ 3999.9995799999997 │
│     1 │           1 │               3500 │
│     1 │           1 │               3000 │
└───────┴─────────────┴────────────────────┘
```

On 20.8.7.15 it returns:

```
SELECT 
    depth,
    currency_id,
    SUM(coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.amount) / dictGet('currency', 'divider', toUInt64(1)) AS `_aq.amount`
FROM coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99
WHERE (coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.depth <= 3) AND (coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.depth > 0)
GROUP BY 
    coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.transfer_from_bin,
    coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.transfer_to_bin,
    coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.currency_id,
    depth
ORDER BY 
    depth ASC,
    `_aq.amount` DESC
LIMIT 10

┌─depth─┬─currency_id─┬─────────_aq.amount─┐
│     1 │           1 │               3000 │
│     1 │           1 │               3500 │
│     1 │           1 │               3000 │
│     1 │           1 │              16001 │
│     1 │           1 │     4000.099948955 │
│     1 │           1 │               7500 │
│     1 │           1 │               8000 │
│     1 │           1 │ 3999.9995799999997 │
│     1 │           1 │ 12388.999999999998 │
│     1 │           1 │                0.1 │
└───────┴─────────────┴────────────────────┘

10 rows in set. Elapsed: 0.002 sec. 


```

**Comment by qoega** [comment]
Does it still reproduce if you filter rows by `HAVING depth=1`? And you can remove `currency_id` column - it should be irrelevant.

**Comment by astudnev** [comment]
Yes, it is the same with HAVING depth=1 and without currency_id:

```
SELECT 
    depth,
    SUM(coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.amount) / dictGet('currency', 'divider', toUInt64(1)) AS `_aq.amount`
FROM coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99
WHERE (coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.depth <= 3) AND (coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.depth > 0)
GROUP BY 
    coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.transfer_from_bin,
    coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.transfer_to_bin,
    depth
HAVING depth = 1
ORDER BY 
    depth ASC,
    `_aq.amount` DESC
LIMIT 10

┌─depth─┬─────────_aq.amount─┐
│     1 │ 12388.999999999998 │
│     1 │               7000 │
│     1 │              16001 │
│     1 │               7500 │
│     1 │               3500 │
│     1 │ 3999.9995799999997 │
│     1 │     4000.099948955 │
│     1 │               4000 │
│     1 │                0.1 │
│     1 │               3000 │
└───────┴────────────────────┘

10 rows in set. Elapsed: 0.002 sec. 
```

**Comment by nikitamikhaylov** [comment]
Which type do depth and _aq.amount have? Could you please execute EXPLAIN SYNTAX <your query> in 20.8?

**Comment by qoega** [comment]
And what if there is no more `depth` in `ORDER BY`?

**Comment by astudnev** [comment]
Without depth in order by it works just fine:

```
SELECT 
    depth,
    SUM(coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.amount) / dictGet('currency', 'divider', toUInt64(1)) AS `_aq.amount`
FROM coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99
WHERE (coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.depth <= 3) AND (coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.depth > 0)
GROUP BY 
    coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.transfer_from_bin,
    coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.transfer_to_bin,
    depth
HAVING depth = 1
ORDER BY `_aq.amount` DESC
LIMIT 10

┌─depth─┬─────────_aq.amount─┐
│     1 │              16001 │
│     1 │ 12388.999999999998 │
│     1 │               8000 │
│     1 │               7500 │
│     1 │               7000 │
│     1 │     4000.099948955 │
│     1 │               4000 │
│     1 │ 3999.9995799999997 │
│     1 │               3500 │
│     1 │               3000 │
└───────┴────────────────────┘

10 rows in set. Elapsed: 0.002 sec. 

```

**Comment by den-crane** [comment]
What if  ORDER BY identity( depth ) ASC,  `_aq.amount` DESC
What if  ORDER BY cast( depth as String ) ASC,  `_aq.amount` DESC

**Comment by astudnev** [comment]
```
SELECT 
    depth,
    SUM(coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.amount) / dictGet('currency', 'divider', toUInt64(1)) AS `_aq.amount`
FROM coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99
WHERE (coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.depth <= 3) AND (coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.depth > 0)
GROUP BY 
    coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.transfer_from_bin,
    coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.transfer_to_bin,
    depth
HAVING depth = 1
ORDER BY 
    identity(depth) ASC,
    `_aq.amount` DESC
LIMIT 10

┌─depth─┬─────────_aq.amount─┐
│     1 │              16001 │
│     1 │ 12388.999999999998 │
│     1 │               8000 │
│     1 │               7500 │
│     1 │               7000 │
│     1 │     4000.099948955 │
│     1 │               4000 │
│     1 │ 3999.9995799999997 │
│     1 │               3500 │
│     1 │               3000 │
└───────┴────────────────────┘

10 rows in set. Elapsed: 0.002 sec. 
```

**Comment by astudnev** [comment]
```
SELECT 
    depth,
    SUM(coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.amount) / dictGet('currency', 'divider', toUInt64(1)) AS `_aq.amount`
FROM coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99
WHERE (coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.depth <= 3) AND (coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.depth > 0)
GROUP BY 
    coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.transfer_from_bin,
    coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99.transfer_to_bin,
    depth
HAVING depth = 1
ORDER BY 
    CAST(depth, 'String') ASC,
    `_aq.amount` DESC
LIMIT 10

┌─depth─┬─────────_aq.amount─┐
│     1 │              16001 │
│     1 │ 12388.999999999998 │
│     1 │               8000 │
│     1 │               7500 │
│     1 │               7000 │
│     1 │     4000.099948955 │
│     1 │               4000 │
│     1 │ 3999.9995799999997 │
│     1 │               3500 │
│     1 │               3000 │
└───────┴────────────────────┘

10 rows in set. Elapsed: 0.002 sec. 
```

**Comment by den-crane** [comment]
What is the type of `depth` column?
What is the table `order by` ?

**Comment by astudnev** [comment]
CREATE TABLE coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99
(
    `depth` UInt32,
    `currency_id` UInt32,
    `block` UInt32,
    `tx_time` DateTime,
    `tx_hash_bin` String,
    `transfer_from_bin` String,
    `transfer_to_bin` String,
    `tx_value` Float64,
    `amount` Float64
)
ENGINE = TinyLog

**Comment by den-crane** [comment]
Probably you can share this table because it's open data ?

clickhouse-client -q "select * from coinpath.txs_ethereum_outbound_4c4dce9598b8e2addcc99 format Native" | gzip > test.native.gz

**Comment by astudnev** [comment]
[Uploading test.native.gz…](Here it is)

**Comment by astudnev** [comment]
[test.native.gz](https://github.com/ClickHouse/ClickHouse/files/5674284/test.native.gz)

**Comment by den-crane** [comment]
I've reproduced but only with dict.

```sql
create table currency_t (id UInt64, divider Float64 ) engine=MergeTree order by tuple();
insert into currency_t values (1, 1000000000000000000)
CREATE DICTIONARY currency (id UInt64, divider Float64)
PRIMARY KEY id SOURCE(CLICKHOUSE(HOST 'localhost' PORT 9000 
TABLE currency_t DB 'default' USER 'default')) 
LIFETIME(MIN 0 MAX 0) LAYOUT(HASHED());

```

**Comment by den-crane** [comment]
even without a table

```sql

SELECT
    d,
    SUM(am) / dictGet('dw.currency', 'divider', toUInt64(1)) AS a
FROM (select toUInt32(1) d,'1' f ,'2' t,3e21 am union all select 1,'1','3',1e18)
GROUP BY d, f,t
ORDER BY (d) , a 

┌─d─┬────a─┐
│ 1 │ 3000 │
│ 1 │    1 │
└───┴──────┘
```

**Comment by CurtizJ** [comment]
Should be fixed in #17471. Just need to  backport to 20.8.
Workaround for now: `set optimize_redundant_functions_in_order_by = 0`

```sql
EXPLAIN SYNTAX
SELECT
    d,
    SUM(am) / dictGet('default.currency', 'divider', toUInt64(1)) AS a
FROM 
(
    SELECT
        toUInt32(1) AS d,
        '1' AS f,
        '2' AS t,
        3e21 AS am
    UNION ALL
    SELECT
        1,
        '1',
        '3',
        1000000000000000000.
)
GROUP BY
    d,
    f,
    t
ORDER BY
    d ASC,
    a ASC
SETTINGS optimize_redundant_functions_in_order_by = 1

Query id: 6c86d062-6641-406e-bdfc-6fd4def40767

┌─explain────────────────────────────────────────────────────────────────┐
│ SELECT                                                                 │
│     d,                                                                 │
│     SUM(am) / dictGet('default.currency', 'divider', toUInt64(1)) AS a │
│ FROM                                                                   │
│ (                                                                      │
│     SELECT                                                             │
│         toUInt32(1) AS d,                                              │
│         '1' AS f,                                                      │
│         '2' AS t,                                                      │
│         3e21 AS am                                                     │
│     UNION ALL                                                          │
│     SELECT                                                             │
│         1,                                                             │
│         '1',                                                           │
│         '3',                                                           │
│         1000000000000000000.                                           │
│ )                                                                      │
│ GROUP BY                                                               │
│     d,                                                                 │
│     f,                                                                 │
│     t                                                                  │
│ ORDER BY d ASC                                                         │
│ SETTINGS optimize_redundant_functions_in_order_by = 1                  │
└────────────────────────────────────────────────────────────────────────┘
```

```sql
EXPLAIN SYNTAX
SELECT
    d,
    SUM(am) / dictGet('default.currency', 'divider', toUInt64(1)) AS a
FROM 
(
    SELECT
        toUInt32(1) AS d,
        '1' AS f,
        '2' AS t,
        3e21 AS am
    UNION ALL
    SELECT
        1,
        '1',
        '3',
        1000000000000000000.
)
GROUP BY
    d,
    f,
    t
ORDER BY
    d ASC,
    a ASC
SETTINGS optimize_redundant_functions_in_order_by = 0

Query id: fd11152d-b19b-489e-9979-f65d7a7966b7

┌─explain────────────────────────────────────────────────────────────────┐
│ SELECT                                                                 │
│     d,                                                                 │
│     SUM(am) / dictGet('default.currency', 'divider', toUInt64(1)) AS a │
│ FROM                                                                   │
│ (                                                                      │
│     SELECT                                                             │
│         toUInt32(1) AS d,                                              │
│         '1' AS f,                                                      │
│         '2' AS t,                                                      │
│         3e21 AS am                                                     │
│     UNION ALL                                                          │
│     SELECT                                                             │
│         1,                                                             │
│         '1',                                                           │
│         '3',                                                           │
│         1000000000000000000.                                           │
│ )                                                                      │
│ GROUP BY                                                               │
│     d,                                                                 │
│     f,                                                                 │
│     t                                                                  │
│ ORDER BY                                                               │
│     d ASC,                                                             │
│     a ASC                                                              │
│ SETTINGS optimize_redundant_functions_in_order_by = 0                  │
└────────────────────────────────────────────────────────────────────────┘
```

**Comment by den-crane** [comment]
yes
```sql
set optimize_redundant_functions_in_order_by=0;

SELECT     d,     SUM(am) / dictGet('dw.currency', 'divider', toUInt64(1)) AS a 
FROM (select toUInt32(1) d,'1' f ,'2' t,3e21 am union all select 1,'1','3',1e18) GROUP BY d,f,t ORDER BY d , a

┌─d─┬────a─┐
│ 1 │    1 │
│ 1 │ 3000 │
└───┴──────┘

set optimize_redundant_functions_in_order_by=1;

┌─d─┬────a─┐
│ 1 │ 3000 │
│ 1 │    1 │
└───┴──────┘
```

**Comment by astudnev** [comment]
Yes
set optimize_redundant_functions_in_order_by=0;
solved this for my query as well

**Comment by CurtizJ** [comment]
It turned out, that it's already backported and fixed since 20.8.8.
